test/rate_limiter_test.cc

```diff
@@ -191,43 +191,80 @@ TEST_F(RateLimiterTest, DistributionSamplingRateLimiterImplTest) {
+// relying on simulated time and a mock discrete numberic distribution sampler.
+class DistributionSamplingRateLimiterTest : public RateLimiterTest {
+public:
+  DistributionSamplingRateLimiterTest()
+      : mock_rate_limiter_(std::make_unique<NiceMock<MockRateLimiter>>()),
+        unsafe_mock_rate_limiter_(*mock_rate_limiter_),
+        unsafe_discrete_numeric_distribution_sampler_(new MockDiscreteNumericDistributionSampler()),
+        rate_limiter_(std::make_unique<DistributionSamplingRateLimiterImpl>(
+            std::unique_ptr<DiscreteNumericDistributionSampler>(
+                unsafe_discrete_numeric_distribution_sampler_),
+            std::move(mock_rate_limiter_))) {
+    EXPECT_CALL(unsafe_mock_rate_limiter_, timeSource)
+        .Times(AtLeast(0))
+        .WillRepeatedly(ReturnRef(time_system_));
+  }
 
-  EXPECT_CALL(unsafe_mock_rate_limiter, tryAcquireOne)
-      .Times(AtLeast(1))
-      .WillRepeatedly(Return(true));
-  EXPECT_CALL(unsafe_mock_rate_limiter, releaseOne).Times(1);
-  EXPECT_CALL(*unsafe_discrete_numeric_distribution_sampler, getValue)
-      .Times(3)
+  std::unique_ptr<NiceMock<MockRateLimiter>> mock_rate_limiter_;
+  MockRateLimiter& unsafe_mock_rate_limiter_;
+  Envoy::Event::SimulatedTimeSystem time_system_;
+  MockDiscreteNumericDistributionSampler* unsafe_discrete_numeric_distribution_sampler_;
+  RateLimiterPtr rate_limiter_;
+};
+
+TEST_F(DistributionSamplingRateLimiterTest, SingleAcquisition) {
+  EXPECT_CALL(unsafe_mock_rate_limiter_, tryAcquireOne)
+      .WillOnce(Return(true))
+      .WillOnce(Return(false))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*unsafe_discrete_numeric_distribution_sampler_, getValue).WillOnce(Return(1));
+  // The distribution first yields a 1 ns offset. So we don't expect to be green lighted.
+  EXPECT_FALSE(rate_limiter_->tryAcquireOne());
+  time_system_.advanceTimeWait(1ns);
+  EXPECT_TRUE(rate_limiter_->tryAcquireOne());
+  EXPECT_FALSE(rate_limiter_->tryAcquireOne());
+}
+
+TEST_F(DistributionSamplingRateLimiterTest, QueuedAcquisition) {
+  EXPECT_CALL(unsafe_mock_rate_limiter_, tryAcquireOne)
+      .WillOnce(Return(true))
+      .WillOnce(Return(true))
+      .WillOnce(Return(false))
+      .WillOnce(Return(false))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*unsafe_discrete_numeric_distribution_sampler_, getValue)
       .WillOnce(Return(1))
-      .WillOnce(Return(0))
       .WillOnce(Return(1));
-
   // The distribution first yields a 1 ns offset. So we don't expect to be green lighted.
-  EXPECT_FALSE(rate_limiter->tryAcquireOne());
-  time_system.advanceTimeWait(1ns);
-  EXPECT_TRUE(rate_limiter->tryAcquireOne());
-  // We expect releaseOne to be propagated.
-  rate_limiter->releaseOne();
-  // The distribution will yield an offset of 0ns, we expect success.
-  EXPECT_TRUE(rate_limiter->tryAcquireOne());
+  EXPECT_FALSE(rate_limiter_->tryAcquireOne());
+  EXPECT_FALSE(rate_limiter_->tryAcquireOne());
```

Comment:
I think the test makes sense, but could you add multiple comments to clarify the scenario, which seems to be:
1. The inner rate limiter releases 2 requests only (the true, true, false, false, false)
2. The 2 requests are delayed until 1ns has elapsed (the 1, 1)
3. After the 2 delayed requests are released, no further requests are released because the underlying rate limiter released nothing further

Reply:
Done